indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/CostBasedAutoScaler.java

```diff
@@ -0,0 +1,310 @@
+      String dataSource,
+      CostBasedAutoScalerConfig config,
+      SupervisorSpec spec,
+      ServiceEmitter emitter
+  )
+  {
+    this.dataSource = dataSource;
+    this.config = config;
+    this.spec = spec;
+    this.supervisor = supervisor;
+    this.emitter = emitter;
+
+    final String supervisorId = StringUtils.format("Supervisor-%s", dataSource);
+
+    this.currentMetrics = new AtomicReference<>(null);
+    this.costFunction = new WeightedCostFunction();
+
+    this.metricsCollectionExec = Execs.scheduledSingleThreaded(
+        StringUtils.encodeForFormat(supervisorId) + "-CostBasedMetrics-%d"
+    );
+    this.scalingDecisionExec = Execs.scheduledSingleThreaded(
+        StringUtils.encodeForFormat(supervisorId) + "-CostBasedScaling-%d"
+    );
+
+    this.metricBuilder = ServiceMetricEvent.builder()
+                                           .setDimension(DruidMetrics.DATASOURCE, dataSource)
+                                           .setDimension(
+                                               DruidMetrics.STREAM,
+                                               this.supervisor.getIoConfig().getStream()
+                                           );
+  }
+
+  @Override
+  public void start()
+  {
+    Callable<Integer> scaleAction = () -> computeOptimalTaskCount(currentMetrics);
+    Runnable onSuccessfulScale = () -> currentMetrics.set(null);
+
+    metricsCollectionExec.scheduleAtFixedRate(
+        collectMetrics(),
+        config.getMetricsCollectionIntervalMillis(),
+        config.getMetricsCollectionIntervalMillis(),
+        TimeUnit.MILLISECONDS
+    );
+
+    scalingDecisionExec.scheduleAtFixedRate(
+        supervisor.buildDynamicAllocationTask(scaleAction, onSuccessfulScale, emitter),
+        config.getScaleActionStartDelayMillis(),
+        config.getScaleActionPeriodMillis(),
+        TimeUnit.MILLISECONDS
+    );
+
+    log.info(
+        "CostBasedAutoScaler started for dataSource [%s]: collecting metrics every [%d]ms, "
+        + "evaluating scaling every [%d]ms",
+        dataSource,
+        config.getMetricsCollectionIntervalMillis(),
+        config.getScaleActionPeriodMillis()
+    );
+  }
+
+  @Override
+  public void stop()
+  {
+    scalingDecisionExec.shutdownNow();
+    metricsCollectionExec.shutdownNow();
+    log.info("CostBasedAutoScaler stopped for dataSource [%s]", dataSource);
+  }
+
+  @Override
+  public void reset()
+  {
+    currentMetrics.set(null);
+  }
+
+  private Runnable collectMetrics()
+  {
+    return () -> {
+      if (spec.isSuspended()) {
+        log.debug("Supervisor [%s] is suspended, skipping metrics collection", dataSource);
+        return;
+      }
+
+      final LagStats lagStats = supervisor.computeLagStats();
+      if (lagStats == null) {
+        log.debug("Lag stats unavailable for dataSource [%s], skipping collection", dataSource);
+        return;
+      }
+
+      final int currentTaskCount = supervisor.getActiveTaskGroupsCount();
```

Comment:
I see that lag-based auto-scaler does this too, but I am not sure if this is correct.
This only returns the number of tasks that are actively reading. There may be tasks which have already moved to publishing. I think the more accurate measure would be the actual `ioConfig.taskCount` itself.

I will need to double check this though. Could you please try to check as well?